firewood/src/merkle/proof.rs

```diff
@@ -258,261 +257,146 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
     /// necessary nodes will be resolved and leave the remaining as hashnode.
     ///
     /// The given edge proof is allowed to be an existent or non-existent proof.
-    fn proof_to_path<KV: AsRef<[u8]>, S: ShaleStore<Node> + Send + Sync, T: BinarySerde>(
+    fn proof_to_path<K: AsRef<[u8]>, S: ShaleStore<Node> + Send + Sync, T: BinarySerde>(
         &self,
-        key: KV,
+        key: K,
         root_hash: [u8; 32],
         merkle_setup: &mut MerkleSetup<S, T>,
         allow_non_existent_node: bool,
     ) -> Result<Option<Vec<u8>>, ProofError> {
         // Start with the sentinel root
-        let root = merkle_setup.get_root();
+        let sentinel = merkle_setup.get_sentinel();
         let merkle = merkle_setup.get_merkle_mut();
-        let mut u_ref = merkle.get_node(root).map_err(|_| ProofError::NoSuchNode)?;
+        let mut parent_node_ref = merkle
+            .get_node(sentinel)
+            .map_err(|_| ProofError::NoSuchNode)?;
 
-        let mut chunks = Vec::new();
-        chunks.extend(key.as_ref().iter().copied().flat_map(to_nibble_array));
+        let mut key_nibbles = Nibbles::<0>::new(key.as_ref()).into_iter();
 
-        let mut cur_key: &[u8] = &chunks;
-        let mut cur_hash = root_hash;
+        let mut child_hash = root_hash;
         let proofs_map = &self.0;
-        let mut key_index = 0;
-        let mut branch_index = 0;
-
-        loop {
-            let cur_proof = proofs_map
-                .get(&cur_hash)
-                .ok_or(ProofError::ProofNodeMissing)?;
-            // TODO(Hao): (Optimization) If a node is already decode we don't need to decode again.
-            let (mut chd_ptr, sub_proof, size) =
-                self.decode_node(merkle, cur_key, cur_proof.as_ref(), false)?;
+        let mut child_index = 0;
```

Comment:
0 is used for the sentinel node because the root-node is always at the 0th child of the sentinel; then, this is only set and used for branch nodes after that. 

The reason we have a variable here is that we have to use the nibble from the iterator twice, once to get the encoded-child and then again set child-address on the parent. 

Where does `(u64, Vec)` come from here and what would it be used for? Is that the encoded-node? And did you mean `u8`? I'm a little confused with what you're suggesting, but I agree that `child_index` is a little awkward here.

Reply:
Sorry, I wasn't super clear. I'm suggesting that you preserve the iterator instead of the offset if you're on a branch node and are iterating that branch node. Then, to get the next child, all you have to do is advance that iterator, possibly using `as_ref()` to avoid terminating the iterator or `clone()` if you want to just look at it and not advance it.

This would let you do some additional simplifications like if the value is in the branch node, just initialize but don't search the iterator. It would also prevent you from going backwards as was shown in the recent StreamIterator bug.